pkg/controller/master/addon/addon.go

```diff
@@ -39,290 +38,528 @@ var (
 type Handler struct {
 	helm  ctlhelmv1.HelmChartController
 	addon ctlharvesterv1.AddonController
-	job   ctlcorev1.JobController
+	job   ctlbatchv1.JobController
 	app   ctlappsv1.AppController
+	pv    ctlcorev1.PersistentVolumeController
+	pvc   ctlcorev1.PersistentVolumeClaimController
 }
 
 func Register(ctx context.Context, management *config.Management, opts config.Options) error {
 	addonController := management.HarvesterFactory.Harvesterhci().V1beta1().Addon()
 	helmController := management.HelmFactory.Helm().V1().HelmChart()
 	jobController := management.BatchFactory.Batch().V1().Job()
 	appController := management.CatalogFactory.Catalog().V1().App()
+	pvController := management.CoreFactory.Core().V1().PersistentVolume()
+	pvcController := management.CoreFactory.Core().V1().PersistentVolumeClaim()
 	h := &Handler{
 		helm:  helmController,
 		addon: addonController,
 		job:   jobController,
 		app:   appController,
+		pv:    pvController,
+		pvc:   pvcController,
 	}
 
-	addonController.OnChange(ctx, "deploy-addon", h.OnAddonChange)
-	addonController.OnChange(ctx, "monitor-addon", h.MonitorAddon)
-	addonController.OnChange(ctx, "monitor-changes", h.MonitorChanges)
+	addonController.OnChange(ctx, "monitor-addon-per-status", h.MonitorAddonPerStatus)
+	addonController.OnChange(ctx, "monitor-addon-rancher-monitoring", h.MonitorAddonRancherMonitoring)
 	appController.OnChange(ctx, "monitor-apps-catalog", h.PatchApps)
 	relatedresource.Watch(ctx, "watch-helmcharts", h.ReconcileHelmChartOwners, addonController, helmController)
 	return nil
 }
 
-// MonitorChanges will trigger updates to helm chart based on changes in values
-func (h *Handler) MonitorChanges(key string, aObj *harvesterv1.Addon) (*harvesterv1.Addon, error) {
+// MonitorAddonPerStatus will track the OnChange on addon, route actions per addon status
+func (h *Handler) MonitorAddonPerStatus(key string, aObj *harvesterv1.Addon) (*harvesterv1.Addon, error) {
 	if aObj == nil || aObj.DeletionTimestamp != nil {
 		return nil, nil
 	}
 
-	a := aObj.DeepCopy()
-	if a.Status.Status != harvesterv1.AddonDeployed && a.Status.Status != harvesterv1.AddonFailed {
-		return a, nil
+	switch aObj.Status.Status {
+	// finish the onging enable
+	case harvesterv1.AddonEnabled:
```

Comment:
based on the workflow of `processAddonEnabling` we should be checking the status `harvesterv1.AddonEnabling`